docs/guides/transactions.md

```diff
@@ -1,6 +1,56 @@
 # Using transactions on the lakeFS file system
 
-A transaction defers file transfers and versioning operations to a queue, which is unwound sequentially on completion.
-Transactions are thread-safe and atomic, meaning that a single failure during any transaction function causes the entire transaction to abort.
+In addition to file operations, you can carry out versioning operations in your Python code using file system *transactions*.
 
-TODO: Add content
+A transaction is basically a context manager that collects all file uploads, defers them, and executes the uploads on completion of the transaction.
+They are an "all or nothing" proposition: If an error occurs during the transaction, none of the queued files are uploaded.
+
+The main features of the lakeFS file system transaction are:
+
+## Thread safety
+
+Through its use of `collections.deque` as a store for uploads, upload queueing and file transfers are thread-safe.
+
+## Atomicity
+
+If an exception occurs anywhere during the transaction, all queued file uploads are discarded:
+
+```python
+from lakefs_spec import LakeFSFileSystem
+
+fs = LakeFSFileSystem()
+
+with fs.transaction as tx:
+    fs.put_file("my-file.txt", "repo/main/my-file.txt")
+    tx.commit("repo", "main", message="Add my-file.txt")
+    raise ValueError("oops!")
+```
+
+The above code will not produce a commit on `main`, since the `ValueError` prompts a discard of the full upload queue. 
+
+## Versioning helpers
+
+The lakeFS file system's transaction is the intended place for conducting versioning operations between file transfers.
+The following is an example of file uploads with commit creations, with a tag being applied at the end.
+
+```python
+from lakefs_spec import LakeFSFileSystem
+
+fs = LakeFSFileSystem()
+
+with fs.transaction as tx:
+    fs.put_file("train-data.txt", "repo/main/train-data.txt")
+    tx.commit("repo", "main", message="Add training data")
+    fs.put_file("test-data.txt", "repo/main/test-data.txt")
+    sha = tx.commit("repo", "main", message="Add test data")
+    tx.tag("repo", sha, tag="My train-test split")
+```
+
+!!! Warning
+
+    All of the operations above are deferred, and their results are not available until completion of the transaction.
+    For example, the `sha` return value of `tx.commit` will be a placeholder for the actual commit SHA computed by the lakeFS
+    server on commit creation.
+
+    While you can use some values (branch/tag names) returned by transaction versioning helpers, we strongly
+    advise not to reuse values computed in transactions since they might result in unexpected behavior.
```

Comment:
> reuse

Not quite sure what this is referring to

Reply:
This is referring to the fact that `sha = tx.commit(...)` will return a `Placeholder[Commit]`, which will not work if passed to a non-transaction function later. I thought this was worth mentioning, since the following seems quite likely to happen eventually:

```python
with fs.transaction as tx:
    tx.put_file("my-file.txt", "repo/branch/my-file.txt")
    sha = tx.commit("repo", "branch", message="Add file xyz")

...
fs.get_file("repo/{sha}/file.txt", "my-new-file.txt")  # <- bang: sha is Placeholder[Commit]
```

What do you think?

Reply:
Ah, that makes total sense (also, that issue would be addressed if we turn the `Placeholder` into a transparent wrapper - at least for accessing the value outside of the current transaction). I'd include the example you gave in the docs for clarity and maybe change "reuse" to "reuse outside of the transaction" (because that is the unexpected bit, I'd say)